### PR TITLE
Validate Unique Saved GeoPolygon Names (Geospatial)

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

An error message will be shown in the UI if the user tries to save a GeoPolygon with the same name as one that already exists in the domain:
Screenshot from 2024-07-03 11-05-45

## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-3718).

A straightforward PR where validation is being added to the POST Django view for GeoPolygons to ensure that the given GeoPolygon name is unique. This validation is case-insensitive, so even if the same name with different capitalization is used, an error will still be shown.

Originally I wanted to do this validation in JS, however I ended up implementing the validation in the back-end to prevent the following bug:
If a user saves a new polygon and then changes the report filters, the newly saved polygon will not be available on the front-end and so the user will be able to save a polygon with the same name. This is because we fetch the list of saved polygons from `initialPageData` on page load or when changing report filters. Since this list comes from `initialPageData` which gets set in the template on page load, changing report filters means we will have a list of saved polygons which does not include the newly saved polygon.

## Feature Flag

`GEOSPATIAL`

### Safety story

- Local testing
- Unit tests

This change will only affect domains that have the Geospatial feature flag enabled.

### Automated test coverage

Unit tests have been implemented to verify all validation checks implemented in the `GeoPolygonListView` view.

### QA Plan

No QA planned.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/geospatial/static/geospatial/js/models.js [hunk 1]
```diff
@@ -771,12 +771,8 @@ hqDefine('geospatial/js/models', [
...
             let data = self.mapObj.drawControls.getAll();
             if (data.features.length) {
-                let name = window.prompt(gettext("Name of the Area"));
-                if (name === null) {
-                    return;
-                }
-                if (name === '') {
-                    alertUser.alert_user(gettext("Please enter the name for the area!"), 'warning', false, true);
+                const name = window.prompt(gettext("Name of the Area"));
+                if (!validateSavedPolygonName(name)) {
```
Comment (ajeety4) [inline corehq/apps/geospatial/static/geospatial/js/models.js:775]: :+1:

File: corehq/apps/geospatial/views.py [hunk 2]
```diff
@@ -78,17 +78,17 @@ def post(self, request, *args, **kwargs):
         try:
             geo_json = json.loads(request.body).get('geo_json', None)
         except json.decoder.JSONDecodeError:
-            raise HttpResponseBadRequest(
+            return HttpResponseBadRequest(
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/views.py:81]: The `HttpResponseBadRequest` class is a subclass of `HttpResponse` and so, isn't actually an `Exception` class at all. Raising it resulted in an uncaught `exceptions must derive from BaseException` exception being raised.

File: corehq/apps/geospatial/tests/test_views.py [hunk 3]
```diff
@@ -453,6 +453,42 @@ def test_save_polygon(self):
             del feature['id']
         self.assertEqual(saved_polygons[0].geo_json, geo_json_data)
 
+    def _validate_error_message(self, response, message):
```
Comment (ajeety4) [inline corehq/apps/geospatial/tests/test_views.py:456]: nit: maybe rename to `_assert_error_message` ?

**Review by ajeety4 — APPROVED** [review]
Looks clear to me. Nice refactor along the way :+1:

File: corehq/apps/geospatial/tests/test_views.py (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/tests/test_views.py]: That sounds good, renamed in [114b4d9](https://github.com/dimagi/commcare-hq/pull/34849/commits/114b4d9cf89e2867788bd3bb527c96045694aec2)

**Comment by ajeety4** [comment]
> Originally I wanted to do this validation in JS, however I ended up implementing the validation in the back-end to prevent the following bug:
If a user saves a new polygon and then changes the report filters, the newly saved polygon will not be available on the front-end and so the user will be able to save a polygon with the same name. This is because we fetch the list of saved polygons from initialPageData on page load or when changing report filters. Since this list comes from initialPageData which gets set in the template on page load, changing report filters means we will have a list of saved polygons which does not include the newly saved polygon.

Great catch w.r.t bug.
One thing I wanted to say is having this kind of validation in the backend in case of REST APIs seems like the right place to me. It is possible to bypass the JS validation and invoke the API directly.

**Comment by zandre-eng** [comment]
> One thing I wanted to say is having this kind of validation in the backend in case of REST APIs seems like the right place to me. It is possible to bypass the JS validation and invoke the API directly.

@ajeety4 That is a good point. In general one should never trust any data coming from a user's browser, as this can always be tweaked or modified even if there is UI validation. 

But it does feel a little weird having some of our validation lie in JS and some in the back-end view. An open question I have is whether we should rather just move all the front-end validation to the view. That way we can keep the validation central and ensure we have the proper validation for when this endpoint is invoked through an API.

**Comment by ajeety4** [comment]
> > One thing I wanted to say is having this kind of validation in the backend in case of REST APIs seems like the right place to me. It is possible to bypass the JS validation and invoke the API directly.
> 
> @ajeety4 That is a good point. In general one should never trust any data coming from a user's browser, as this can always be tweaked or modified even if there is UI validation.
Agreed
 
> But it does feel a little weird having some of our validation lie in JS and some in the back-end view. An open question I have is whether we should rather just move all the front-end validation to the view. That way we can keep the validation central and ensure we have the proper validation for when this endpoint is invoked through an API.

Good point.  
My take on this as below:
- For REST APIs or a platform with separate client/server apps, I think we must have all the validations in the backend. So I completely agree with you on moving all the validations to the views in this scenario.
- For JS, I think it is good to have the same validation wherever possible (completely optional) for e.g. same name validation . This could improve user experience and reduce network calls.

I liked the idea of making the API robust in terms of validation, maybe create a separate ticket for this if it feels worth to do ?

**Review by ajeety4 — APPROVED** [review]

**Comment by zandre-eng** [comment]
> My take on this as below:
For REST APIs or a platform with separate client/server apps, I think we must have all the validations in the backend. So I completely agree with you on moving all the validations to the views in this scenario.
For JS, I think it is good to have the same validation wherever possible (completely optional) for e.g. same name validation . This could improve user experience and reduce network calls.
I liked the idea of making the API robust in terms of validation, maybe create a separate ticket for this if it feels worth to do ?

@ajeety4 That sounds right, I like the idea of keeping the validation in JS still so that we can avoid API calls where possible. I'll create a separate ticket so that we can make the view validation more robust.

**Pull request merged**